src/format.rs

```diff
@@ -1,97 +1,247 @@
-#[inline(always)]
-fn format_timestamp(buf: &mut Vec<u8>) {
+fn format_timestamp(buf: &mut [u8]) {
     use chrono::{Datelike, Timelike};
 
     let timestamp = chrono::Utc::now();
-    write!(
-        buf,
-        "{:004}-{:02}-{:02}T{:02}:{:02}:{:02}.{:06}Z ",
-        timestamp.year(),
-        timestamp.month(),
-        timestamp.day(),
-        timestamp.hour(),
-        timestamp.minute(),
-        timestamp.second(),
-        timestamp.nanosecond() / 1000,
-    )
-    .unwrap_or_else(|_| unreachable!());
+    let mut itoa = itoa::Buffer::new();
+
+    buf[4..8].copy_from_slice(itoa.format(timestamp.year()).as_bytes());
+    zero_pad2(&mut buf[9..11], itoa.format(timestamp.month()).as_bytes());
+    zero_pad2(&mut buf[12..14], itoa.format(timestamp.day()).as_bytes());
+    zero_pad2(&mut buf[15..17], itoa.format(timestamp.hour()).as_bytes());
+    zero_pad2(&mut buf[18..20], itoa.format(timestamp.minute()).as_bytes());
+    zero_pad2(&mut buf[21..23], itoa.format(timestamp.second()).as_bytes());
+    zero_pad6(
+        &mut buf[24..30],
+        itoa.format(timestamp.nanosecond() / 1000).as_bytes(),
+    );
 }
 
-/// Prints key values in ": key1=value1, key2=value2" format.
-///
-/// # Notes
-///
-/// Prints ": " itself, only when there is at least one key value pair.
-struct KeyValuePrinter<'a>(&'a dyn kv::Source);
-
-impl<'a> fmt::Display for KeyValuePrinter<'a> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        self.0
-            .visit(&mut KeyValueVisitor(true, f))
-            .map_err(|_| fmt::Error)
+#[inline]
+#[cfg(feature = "timestamp")]
+fn zero_pad2(buf: &mut [u8], v: &[u8]) {
+    debug_assert_eq!(buf.len(), 2);
+    if v.len() == 1 {
+        buf[0] = b'0';
+        buf[1] = v[0];
+    } else {
+        buf[0] = v[0];
+        buf[1] = v[1];
     }
 }
 
-struct KeyValueVisitor<'a, 'b>(bool, &'a mut fmt::Formatter<'b>);
+#[inline]
+#[cfg(feature = "timestamp")]
+fn zero_pad6(buf: &mut [u8], v: &[u8]) {
+    debug_assert_eq!(buf.len(), 6);
+    let start = 6 - v.len();
+    for i in 0..=start {
+        buf[i] = b'0';
+    }
+    buf[start..6].copy_from_slice(v);
+}
 
-impl<'a, 'b, 'kvs> kv::Visitor<'kvs> for KeyValueVisitor<'a, 'b> {
+/// Formats key value pairs in the following format: `key="value"`. For example:
+/// `user_name="Thomas" user_id=123 is_admin=true`
+struct KeyValueVisitor<'b>(&'b mut Vec<u8>);
+
+impl<'b, 'kvs> kv::Visitor<'kvs> for KeyValueVisitor<'b> {
     fn visit_pair(&mut self, key: kv::Key<'kvs>, value: kv::Value<'kvs>) -> Result<(), kv::Error> {
-        self.1
-            .write_str(if self.0 { ": " } else { ", " })
-            .and_then(|()| {
-                self.0 = false;
-                write!(self.1, "{}={}", key, value)
-            })
-            .map_err(Into::into)
+        self.0.push(b' ');
+        self.0.extend_from_slice(key.as_str().as_bytes());
+        self.0.push(b'=');
+        if let Some(value) = value.to_borrowed_str() {
```

Comment:
These `to_*` methods aren't really meant for exhaustively visiting the value. So instead of writing something like:

```rust
if let Some(v) = value.to_u64() {
  ..
} else if let Some(v) = value.to_u32() {
  ..
} else if ..
```

you can either use the `kv_unstable_sval` feature along with [`sval` in no-std](https://docs.rs/sval/1.0.0-alpha.4/sval/) or the `kv_unstable_serde` feature along with [`serde`](https://docs.rs/serde) to visit the value. Our documentation is still very sparse so this is really not obvious! If you'd like I can put together an example of what this would look like.

It looks like this should be about the same as `write!(self.1, "{}={:?}", key, value)` though, was there a reason you needed to replace the previous code calling `write!(self.1, "{}={}", key, value)`? Was there something different about the format that meant you had to throw the whole `Debug` format out?

Reply:
> These `to_*` methods aren't really meant for exhaustively visiting the value. So instead of writing something like:

Ah, I didn't get that from the documentation.

> you can either use the `kv_unstable_sval` feature along with [`sval` in no-std](https://docs.rs/sval/1.0.0-alpha.4/sval/) or the `kv_unstable_serde` feature along with [`serde` in std](https://docs.rs/serde) to visit the value. Our documentation is still very sparse so this is really not obvious! If you'd like I can put together an example of what this would look like.

I haven't looked too much as sval, but doesn't that require the structure to implement `sval::value::Value`? Same question for serde. Basically I'm looking the get the primitive type from the `Value`, would it perhaps be possible for valuebag/log to expose `valuabag::internal::Primitive` in some way?

> It looks like this should be about the same as `write!(self.1, "{}={:?}", key, value)` though, was there a reason you needed to replace the previous code calling `write!(self.1, "{}={}", key, value)`? Was there something different about the format that meant you had to throw the whole `Debug` format out?

Mainly I'm looking to optimise logging by using more vectored I/O and reducing the overhead of formatting, e.g. I'm also using the `itoa` crate rather than std library's formatting.

Reply:
The `Value` type will already implement `sval::Value` or `serde::Serialize` so all we need to do is write a `Stream` or `Serializer` that will format the way you want.

When I get a chance over the weekend I’ll put something together to show you.

Reply:
[Here's an example of using `sval` to visit the value](https://github.com/Thomasdezeeuw/std-logger/compare/logfmt...KodrAus:logfmt).

It's come up a few times that users who just want to visit a few primitive datatypes end up chaining `to_*` methods though, so I've also opened https://github.com/rust-lang/log/issues/440 to provide a simple zero-dependency visitor API for `log::kv::Value`s that should cover this case without needing to bring in `sval` directly.

Reply:
@KodrAus thanks for the example. One more question, If I have a `kv::key::Source` created using `serde` can I still use `sval::Stream` for this source? I.e. would the correct `Stream::u64`, etc., functions be called? Or would it use the display implementation?

Reply:
@Thomasdezeeuw It would, yes. One of the goals of the design is that you get the closest possible representation of your value without having to worry about what was passed in in the first place.

So if a value is created using `serde` that serializes as a `u64` or struct or sequence then you'll get an appropriate `u64` or map or sequence from `sval`. You'll also get something that looks like typical `#[derive(Debug)]` if you decide to format it using `{:?}` or `{}`.